### PR TITLE
Add type hints for PyTorch BERT.

### LINKED ISSUE #16059: Add missing type hints
### This issue is part of our **Great Code Cleanup 2022**. If you're interested in helping out, take a look at [this thread](https://twitter.com/carrigmat/status/1502319813510766599), or come [join us on Discord](https://t.co/kS42XBvpWH) and talk with other contributors!

# 🚀 Add missing type hints

Type hints are used inconsistently in the `transformers` repo across both TF and PT models, and it'd be nice to make them a complete, consistent thing for the core models, especially because we want to develop features that depend on them!

### Guide to contributing:

1. Ensure you've read our contributing [guidelines](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) 📜 
2. Claim your architecture(s) in this thread (ensure no one is working on it). It's 100% okay to only take the TensorFlow or PyTorch version of a model, if you're not familiar with both frameworks! It's also okay to claim multiple models and group those changes into a single PR! 🎯 
3. Implement the changes as in https://github.com/huggingface/transformers/pull/16057 or https://github.com/huggingface/transformers/pull/16074 (see the diff on the model architectures for a few examples) 💪 
4. Open the PR and tag me in it. You should run `make fixup` at the end to do a code quality check before your final commit!

### Tips for making your PR

1. The files you need to edit will be in `src/transformers/models/[model_name]/`
2. For TensorFlow, you want the `modeling_tf_[model_name].py` file. For PyTorch, you want the `modeling_[model_name].py` file.
3.  Remember, you **do not** have to cover every class in that file!. The main thing we want to cover is the `call` (for TF) or `forward` (for PT) method for user-facing classes like `TFRobertaForMaskedLM` or `RobertaForSequenceClassification`. It's not necessary to add type hints to layers or base classes like `RobertaModel` or `TFRobertaPreTrainedModel` - these are trickier to write, and generally people do not use those classes as standalone models.
4. If you're unfamiliar with how type hints work, you can read the [Python library documentation on them](https://docs.python.org/3/library/typing.html), but it's probably even easier to just look at another PR that added them. Take a look at the list of changes in the pull requests linked above!
5. The types will usually be obvious - most inputs are `Optional[Union[np.ndarray, tf.Tensor]]` for TF models and `Optional[torch.Tensor]` for PyTorch models, and boolean inputs are `Optional[bool]`. Pay attention to the first input of TF models, though, which is usually `TFModelInputType` - this is because Keras handles that first input in a special way! Other inputs to pay attention to are `past_key_values`, which can vary between models, and also the model output type. For the base model classes like `RobertaModel`, you may have to look at the corresponding `MainLayer` to figure out the right output type! Also, note that the output type may be a tuple if `return_dict` is False, in which case you should specify `Union[Tuple, ...]`. Finally, note that in TF models, `training` is never `None`, so it should be `training: bool` and not `training: Optional[bool]`.
6. Note that some code is copied across our codebase. If you see a line like `# Copied from transformers.models.bert...`, this means that the code is copied from that source, and our scripts will automatically keep that in sync. If you see that, you should not edit the copied method! Instead, edit the original method it's copied from, and run `make fixup` to synchronize that across all the copies. Be sure you installed the development dependencies with `pip install -e ".[dev"]`, as described in the contributor guidelines above, to ensure that the code quality tools in `make fixup` can run.

### How can I find models that need type hints?

I used to maintain a list here, but it got out of date, I'm sorry. Instead, you can use [this Colab notebook](https://colab.research.google.com/drive/1EvZTslb50yfRqIcXjCZFrbod4HrPdA0G?usp=sharing). If you run this, it will show you models in PyTorch or TF that are still missing type hints. Unlike my manually curated lists, it's guaranteed to be up to date - but do double-check that someone else in the thread hasn't claimed a model before you start, because the Colab code will only register type hints after the PR containing them is merged!

### PR DESCRIPTION
**Pull request by coledie**
# What does this PR do?

Add type hints for PyTorch BERT.

Fixes #16059 for PyTorch BERT

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- text models: @ArthurZucker and @younesbelkada
- type hints: @Rocketknight1

### CODE REVIEW

File: src/transformers/models/bert/tokenization_bert.py [hunk 1]
```diff
@@ -171,7 +171,7 @@ class BertTokenizer(PreTrainedTokenizer):
 
             This should likely be deactivated for Japanese (see this
             [issue](https://github.com/huggingface/transformers/issues/328)).
-        strip_accents (`bool`, *optional*):
+        strip_accents (`str`, *optional*):
```
Comment (Rocketknight1) [inline src/transformers/models/bert/tokenization_bert.py:174]: Are you sure this is a `str`? The relevant code is in `super()` but it feels like a `bool` to me.

File: src/transformers/models/bert/tokenization_bert.py [hunk 2]
```diff
@@ -183,17 +183,17 @@ class BertTokenizer(PreTrainedTokenizer):
...
+        do_lower_case: bool = True,
+        do_basic_tokenize: bool = True,
+        never_split: Optional[Iterable[str]] = None,
+        unk_token: str = "[UNK]",
+        sep_token: str = "[SEP]",
+        pad_token: str = "[PAD]",
+        cls_token: str = "[CLS]",
+        mask_token: str = "[MASK]",
+        tokenize_chinese_chars: str = True,
+        strip_accents: Optional[str] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/bert/tokenization_bert.py:196]: Same question here

File: src/transformers/models/bert/tokenization_bert.py [hunk 3]
```diff
@@ -382,20 +382,26 @@ class BasicTokenizer(object):
 
             This should likely be deactivated for Japanese (see this
             [issue](https://github.com/huggingface/transformers/issues/328)).
-        strip_accents (`bool`, *optional*):
+        strip_accents (`str`, *optional*):
```
Comment (Rocketknight1) [inline src/transformers/models/bert/tokenization_bert.py:385]: And here

File: src/transformers/models/bert/tokenization_bert.py [hunk 4]
```diff
@@ -382,20 +382,26 @@ class BasicTokenizer(object):
...
             value for `lowercase` (as in the original BERT).
     """
 
-    def __init__(self, do_lower_case=True, never_split=None, tokenize_chinese_chars=True, strip_accents=None):
+    def __init__(
+        self,
+        do_lower_case: bool = True,
+        never_split: Optional[Iterable[str]] = None,
+        tokenize_chinese_chars: bool = True,
+        strip_accents: Optional[str] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/bert/tokenization_bert.py:395]: And here!

File: src/transformers/models/bert/tokenization_bert_fast.py [hunk 5]
```diff
@@ -206,16 +206,16 @@ class BertTokenizerFast(PreTrainedTokenizerFast):
...
+        vocab_file: Optional[str] = None,
+        tokenizer_file: Optional[str] = None,
+        do_lower_case: bool = True,
+        unk_token: str = "[UNK]",
+        sep_token: str = "[SEP]",
+        pad_token: str = "[PAD]",
+        cls_token: str = "[CLS]",
+        mask_token: str = "[MASK]",
+        tokenize_chinese_chars: bool = True,
+        strip_accents: Optional[str] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/bert/tokenization_bert_fast.py:218]: And here!

**Review by Rocketknight1 — APPROVED** [review]
This looks really good, thank you! I have one question about the `strip_accents` argument, but other than that this looks perfect.

**Comment by Rocketknight1** [comment]
Looks good! The last thing you'll need to do is `pip install transformers[quality]` followed by `make style` in the `transformers` directory. This runs our code formatting tools to make sure everything follows our style guidelines. Once you do that and commit any changes, the tests should pass!

If you run `make style` and you get an error, it may indicate some part of your code that has issues that our code style tools can't correct - if that happens, take a look and try to see what's wrong, and reply here if you can't figure it out!

**Comment by coledie** [comment]
It looks like I've got one last error that I cannot figure out, the relevant line of code appears several times in many files without issue otherwise. Could you help with this?

**Comment by coledie** [comment]
@Rocketknight1

**Comment by Rocketknight1** [comment]
Ah, thanks for the ping! Investigating now

**Comment by Rocketknight1** [comment]
I've investigated and there's an issue in our copy checking code, specifically the `is_copy_consistent` function. This isn't your fault, and I'll need to file a PR to fix it!

(For internal `transformers` reference): The issue is triggered when a function is copied from with a single-line header, and there is a change that causes its header to now be multi-line (e.g. adding type hints and causing `black` to wrap the line). The `is_copy_consistent` function builds a replacement function from the first line of the target function followed by the subsequent lines of the original function, which creates a mangled header if the original header has changed to multi-line:

```python
def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None):
        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
    ) -> List[int]:
```

@sgugger do you want me to make a PR to fix it?

**Comment by sgugger** [comment]
Lol no, fixing this is really not a priority. The copies can be manually updated.

**Comment by coledie** [comment]
@Rocketknight1 Looks like it is working since test failures are unrelated?

**Pull request closed without merging**